pkg/controller/controller.go

```diff
@@ -1267,6 +1289,113 @@ func (c *M3DBController) getParentCluster(pod *corev1.Pod) (*myspec.M3DBCluster,
 	return cluster, nil
 }
 
+// cleanupUnusedStorageForSts removes all PVCs and PVs for the stateful which are not used.
+// This method should be called only when the cluster is stable
+// and there are no ongoing deployments.
+func (c *M3DBController) cleanupUnusedStorageForSts(
+	ctx context.Context,
+	cluster *myspec.M3DBCluster,
+	sts *appsv1.StatefulSet,
+) error {
+	logger := c.logger.With(
+		zap.String("cluster", cluster.Name),
+		zap.String("namespace", cluster.Namespace),
+	)
+
+	lbls := labels.BaseLabels(cluster)
+	lbls[labels.StatefulSet] = sts.Name
+
+	labelSelector := klabels.SelectorFromSet(lbls)
+	pvcs, err := c.pvcLister.PersistentVolumeClaims(cluster.Namespace).List(labelSelector)
+	if err != nil {
+		return fmt.Errorf("listing pvcs: %w", err)
+	}
+
+	pvcsToRemoveByName := make(map[string]*corev1.PersistentVolumeClaim, len(pvcs))
+	for _, pvc := range pvcs {
+		pvcsToRemoveByName[pvc.Name] = pvc
+	}
+
+	// first reclaim the PVCs which should be in use deciding by the parsed pod id
+	// and the replica count of the STS
+	lastPodIDInSTS := int(*sts.Spec.Replicas - 1)
```

Comment:
maybe it would be safier to iterate all pod specs in a statefulset, take PVC name from them and filter out from "all pvc list". After that you will be left with pvcs not belonging to any pods.

Reply:
there is this check below